sparkly/testing.py

```diff
@@ -120,28 +147,132 @@ def tearDown(self):
         """
         if fields:
             actual_df = actual_df.select(*fields)
 
         actual_rows = actual_df.collect() if hasattr(actual_df, 'collect') else actual_df
         actual_data = [row.asDict(recursive=True) for row in actual_rows]
 
-        if ordered:
-            self.assertEqual(actual_data, expected_data)
-        else:
-            try:
-                self.assertCountEqual(actual_data, expected_data)
-            except AttributeError:
-                self.assertItemsEqual(actual_data, expected_data)
+        return self.assertEqual(
+            actual_data,
+            expected_data,
+            ignore_order=not ordered,
+            max_recursion_depth=1,
+            places=places,
+        )
+
+    def assertEqual(self,
+                    first,
+                    second,
+                    msg=None,
+                    ignore_order=False,
+                    max_recursion_depth=None,
+                    places=None,
+                    significant=False,
+                    equal_nan=True):
+        """Assert equal on steroids.
+
+        Extend this classic function signature to work better with
+        comparisons involving dataframes, rows, dictionaries and lists by
+        ignoring the order of mutable nested structures recursively and
+        rounding up floats when necessary.
+
+        Args:
+            first, second, msg: see unittest.TestCase.assertCountEqual
+            ignore_order (bool|False): ignore the order in all structures
+                that are inherently oredered (e.g., lists, mock calls).
+            max_recursion_depth (int|None): if ignore_order is true, do
+                ignore order up to this nesting level (exclusive). Setting
+                this to 0 is same to setting ignore_order=False, 1 means
+                ignore order only at the top level and so on. Default is
+                ignore order arbitrarily deep.
+            places (int|None): Decimal or significant digits to consider
+                during comparisons of floats.
+            significant (bool|False): If set, interpret places as # of
+                significant digits instead of decimal.
+            equal_nan (bool|True): If set, NaNs will compare equal.
+
+        Returns:
+            None, iff the two objects are equal.
+
+        Raises
+            AssertionError, iff the two objects are not equal. See
+            unittest.TestCase.assertEqual for details.
+        """
+        def _create_key(entry):
+            if isinstance(entry, type(mock.call)):
+                _, call_args, call_kwargs = entry
+                entry = list(call_args) + list(call_kwargs.items())
+            if isinstance(entry, collections.Iterable) and not isinstance(entry, six.string_types):
+                return entry is None, tuple(_create_key(x) for x in entry)
+            return entry is None, entry
+
+        def _nicefy(obj, recursion_depth=0):
+            if isinstance(obj, float):
```

Comment:
This function is adding multiple separate improvements to assertEqual, and is a bit difficult to understand as a result.  I'd pull out the logic for floats and T.DataTypes into their own functions that you register using TestCase.addTypeEqualityFunc() (https://docs.python.org/2/library/unittest.html#unittest.TestCase.addTypeEqualityFunc).  The float precision params can be configured using an private instance variable. 

This way, your function is more focused around adding the ability to compare and ignore order in complex datastructures.